### PR TITLE
Add libyaml and pyyaml packages

### PR DESCRIPTION
**Pull request by daoxian**
for https://github.com/iodide-project/pyodide/issues/1171
add PyYAML.
Notice:  We should add Cython into our host python in build-env so as to build pyyaml/ext/_yaml.c file, by running 
`pip3 install cython`

### CODE REVIEW

**Comment by daoxian** [comment]
@hoodmane Would you please help me on this issue?  ci/circleci: build-packages — Your tests failed on **CircleCI**

> Writing /root/repo/packages/.artifacts/lib/python/PyYAML-5.3.1-py3.8.egg-info
> /usr/local/lib/python3.8/distutils/dist.py:274: UserWarning: Unknown distribution option: 'python_requires'
>   warnings.warn(msg)
> /usr/local/lib/python3.8/site-packages/Cython/Compiler/Main.py:369: FutureWarning: Cython directive 'language_level' not set, using 2 for now (Py2). This will change in a later release! File: /root/repo/packages/pyyaml/build/pyyaml-5.3.1/ext/_yaml.pxd
>   tree = Parsing.p_module(s, pxd, full_module_name)
> In file included from ext/_yaml.c:608:
> ext/_yaml.h:2:10: fatal error: 'yaml.h' file not found
> #include <yaml.h>
>          ^~~~~~~~
> 1 error generated.
> shared:ERROR: '/root/repo/emsdk/emsdk/fastcomp/fastcomp/bin/clang -target asmjs-unknown-emscripten -D__EMSCRIPTEN_major__=1 -D__EMSCRIPTEN_minor__=38 -D__EMSCRIPTEN_tiny__=44 -D_LIBCPP_ABI_VERSION=2 -Werror=implicit-function-declaration -Xclang -nostdsysteminc -Xclang -isystem/root/repo/emsdk/emsdk/fastcomp/emscripten/system/include/libcxx -Xclang -isystem/root/repo/emsdk/emsdk/fastcomp/emscripten/system/lib/libcxxabi/include -Xclang -isystem/root/repo/emsdk/emsdk/fastcomp/emscripten/system/include/compat -Xclang -isystem/root/repo/emsdk/emsdk/fastcomp/emscripten/system/include -Xclang -isystem/root/repo/emsdk/emsdk/fastcomp/emscripten/system/include/libc -Xclang -isystem/root/repo/emsdk/emsdk/fastcomp/emscripten/system/lib/libc/musl/arch/emscripten -Xclang -isystem/root/repo/emsdk/emsdk/fastcomp/emscripten/system/local/include -DEMSCRIPTEN -I/src/cpython/installs/python-3.8.2/include/python3.8 -I/src/emsdk/emsdk/fastcomp/emscripten/system/include/libc -I/src/packages/libyaml/build/include -Wno-unused-result -Wsign-compare -DNDEBUG -fwrapv -O3 -Wall -fPIC -DYAML_DECLARE_STATIC=1 -I/root/repo/cpython/installs/python-3.8.2/include/python3.8 -c -mllvm -disable-llvm-optzns -g ext/_yaml.c -Xclang -disable-O0-optnone -Xclang -isystem/root/repo/emsdk/emsdk/fastcomp/emscripten/system/include/SDL -c -o build/temp.linux-x86_64-3.8/ext/_yaml.bc -emit-llvm' failed (1)
> emcc -I/src/cpython/installs/python-3.8.2/include/python3.8 -I/src/emsdk/emsdk/fastcomp/emscripten/system/include/libc -I/src/packages/libyaml/build/include -Wno-unused-result -Wsign-compare -DNDEBUG -g -fwrapv -O3 -Wall -fPIC -DYAML_DECLARE_STATIC=1 -I/root/repo/cpython/installs/python-3.8.2/include/python3.8 -c ext/_yaml.c -o build/temp.linux-x86_64-3.8/ext/_yaml.bc
> Traceback (most recent call last):
>   File "/usr/local/lib/python3.8/runpy.py", line 193, in _run_module_as_main
>     return _run_code(code, main_globals, None,
>   File "/usr/local/lib/python3.8/runpy.py", line 86, in _run_code
>     exec(code, run_globals)
>   File "/root/repo/pyodide_build/__main__.py", line 34, in <module>
>     main()
>   File "/root/repo/pyodide_build/__main__.py", line 28, in main
>     args.func(args)
>   File "/root/repo/pyodide_build/buildpkg.py", line 314, in main
>     build_package(path, args)
>   File "/root/repo/pyodide_build/buildpkg.py", line 254, in build_package
>     compile(path, srcpath, pkg, args)
>   File "/root/repo/pyodide_build/buildpkg.py", line 145, in compile
>     subprocess.run(
>   File "/usr/local/lib/python3.8/subprocess.py", line 512, in run
>     raise CalledProcessError(retcode, process.args,
> subprocess.CalledProcessError: Command '['/usr/local/bin/python3', '-m', 'pyodide_build', 'pywasmcross', '--cflags', ' -I/src/cpython/installs/python-3.8.2/include/python3.8\n-I/src/emsdk/emsdk/fastcomp/emscripten/system/include/libc\n-I/src/packages/libyaml/build/include\n', '--ldflags', "-O2 -s BINARYEN_METHOD='native-wasm' -Werror -s EMULATED_FUNCTION_POINTERS=1 -s EMULATE_FUNCTION_POINTER_CASTS=1 -s SIDE_MODULE=1 -s WASM=1 -s BINARYEN_TRAP_MODE='clamp' --memory-init-file 0 -s LINKABLE=1 -s EXPORT_ALL=1 -L/src/packages/libyaml/build/lib\n", '--target', '/root/repo/cpython/installs/python-3.8.2', '--install-dir', '/root/repo/packages/.artifacts']' returned non-zero exit status 1.
> [2021-01-25 03:07:11] Building package pyyaml...
> [2021-01-25 03:07:14] done building package pyyaml in 3.2 s.

I've added libyaml  in the packages dir, and it's compiled successfully in my local docker building-env, 0.16.1.   It seems that CircleCI didn't build the package libyaml, which causes the file _yaml.h not found.  So how to make CircleCI build the libyaml package?

**Comment by hoodmane** [comment]
I will try to build it locally and see if I get any ideas. @dalcde would be more helpful than me for this, maybe he can take a look at it if he has time.

**Comment by hoodmane** [comment]
For what it's worth, I was able to build this locally as well.

**Comment by hoodmane** [comment]
By the way, there is some minor issue with the configuration: the package is named `pyyaml` but the import name is `yaml`, so in `packages.json` the object `import_name_to_package_name` should have `yaml ==> pyyaml` not `pyyaml ==> pyyaml`. I'm not really sure why this problem arises though, it seems to work automatically for other packages.

**Comment by hoodmane** [comment]
As you said, the problem indeed seems to be that CircleCI attempted to build `pyyaml` before building `libyaml`. Comparing this to other packages with C library dependencies, it _looks_ to me like everything is right. Again, I think @dalcde understands how the dependency resolution works when building packages and should hopefully be able to help.

**Comment by dalcde** [comment]
Your PR is based on a fairly old version of master. I would rebase to latest master before troubleshooting any issues, since the build system of libraries has substantially changed (you include the build script in `meta.yaml` instead of a separate makefile, and it takes care of the downloading etc. for you)

**Comment by daoxian** [comment]
> By the way, there is some minor issue with the configuration: the package is named `pyyaml` but the import name is `yaml`, so in `packages.json` the object `import_name_to_package_name` should have `yaml ==> pyyaml` not `pyyaml ==> pyyaml`. This prevents `pyodide.loadPackagesFromImports` / `pyodide.runPythonAsync` from working correctly. I'm not really sure why this problem arises though, it seems to work automatically for other packages.

Thanks for your tips, but I think once pyodide.loadPackage(['pyyaml']) is OK, we needn't care about if import pyaml or yaml or any else name.

**Comment by daoxian** [comment]
> Your PR is based on a fairly old version of master. I would rebase to latest master before troubleshooting any issues, since the build system of libraries has substantially changed (you include the build script in `meta.yaml` instead of a separate makefile, and it takes care of the downloading etc. for you)

Thank you and I'll try it.

**Comment by hoodmane** [comment]
> Thanks for your tips, but I think once pyodide.loadPackage(['pyyaml']) is OK, we needn't care about if import pyaml or yaml or any else name.

The code that should work but doesn't is `pyodide.runPythonAsync("import yaml")` without first explicitly loading the `pyyaml` package.

File: packages/pyyaml/meta.yaml [hunk 1]
```diff
@@ -0,0 +1,28 @@
...
+  ldflags: |
+    -L/src/packages/libyaml/build/lib
+
+requirements:
+  run:
+    - libyaml
+
+test:
+  imports:
+    - pyyaml
```
Comment (rth) [inline packages/pyyaml/meta.yaml:28]: The issue is that this should be,
```suggestion
    - yaml
```
I think. As it's also used for determining import > package mapping.

Comment (hoodmane) [reply packages/pyyaml/meta.yaml:28]: It's a bit confusing to me that the field which determines import > package mapping is under `test`. I guess the docs say `test/imports` is the "List of imports to test after the package is built." which is taken also the set of python imports to expose that depend on this package.

Presumably having `pyyaml` here would have caused the packages tests to fail on `ImportError` when attempting to import `pyyaml` if the packages were built successfully.

Comment (rth) [reply packages/pyyaml/meta.yaml]: Yes, is currently a bit confusing but we are also trying to follow conda's [meta.yaml spec](https://docs.conda.io/projects/conda-build/en/latest/resources/define-metadata.html) as much as possible. Not sure that changing the name of the top level key would be worth it. Maybe we just need better documentation.

**Pull request closed without merging**